numpyro/contrib/nn/block_neural_arn.py

```diff
@@ -0,0 +1,182 @@
+from jax import ops, random
+from jax.experimental import stax
+from jax.nn import sigmoid, softplus
+from jax.nn.initializers import glorot_uniform, normal, uniform
+import jax.numpy as np
+
+from numpyro.distributions.util import logmatmulexp, vec_to_tril_matrix
+
+
+def BlockMaskedDense(num_blocks, in_factor, out_factor, bias=True, W_init=glorot_uniform()):
+    """
+    Module that implements a linear layer with block matrices with positive diagonal blocks.
+    Moreover, it uses Weight Normalization (https://arxiv.org/abs/1602.07868) for stability.
+
+    :param int num_blocks: Number of block matrices.
+    :param int in_factor: number of rows in each block.
+    :param int out_factor: number of columns in each block.
+    :param W_init: initialization method for the weights.
+    :return: an (`init_fn`, `update_fn`) pair.
+    """
+    input_dim, out_dim = num_blocks * in_factor, num_blocks * out_factor
+    # construct mask_d, mask_o for formula (8) of Ref [1]
+    # Diagonal block mask
+    mask_d = np.identity(num_blocks)[..., None]
+    mask_d = np.tile(mask_d, (1, in_factor, out_factor)).reshape(input_dim, out_dim)
+    # Off-diagonal block mask for lower triangular weight matrix
+    mask_o = vec_to_tril_matrix(np.ones(num_blocks * (num_blocks - 1) // 2), diagonal=-1)[..., None]
```

Comment:
issue: we should create an upper triangular weight matrix
fix: transpose the output of `vec_to_tril_matrix`